### ISSUE TITLE
[mu bug]

### ISSUE THREAD

**Issue by sje30** [djcb/mu#2116]
**Describe the bug**

Pressing 'e' in an *Article* buffer, generates the message:

No attached files found

despite there being attachments that I can view when clicking on their buttons of the Attachments: line in the header.

Investigating this a bit deeper, I see the mime part structure differs to what is expected.  running

    (pp (mu4e~view-gather-mime-parts))

```
((7 #<buffer  *mm*-155504>
     ("application/octet-stream"
     (name . "Important Information Document.pdf"))
    base64 nil
    ("attachment")
    nil nil nil)
 (6 #<buffer  *mm*-298653>
    ("application/octet-stream"
     (name . "Primary IPID.pdf"))
    base64 nil
    ("attachment")
    nil nil nil)
 (5 #<buffer  *mm*-380027>
    ("application/octet-stream"
     (name . "Policy Wording.pdf"))
    base64 nil
    ("attachment")
    nil nil nil)
 (4 #<buffer  *mm*-158383>
    ("application/octet-stream"
     (name . "Policy Pack.pdf"))
    base64 nil
    ("attachment")
    nil nil nil)
 (2 #<buffer  *mm*-984499>
    ("text/plain"
     (charset . "utf-8"))
    base64 nil
    ("inline")
    nil nil nil))
```

where the (attachment) section is empty; compared with a normal message which has the filename element wihtin the attachment.

```
"((4 #<buffer  *mm*-216735>
    (\"application/pdf\"
     (name . \"UKRI-BBSRC_Scientific_Computing_in_the_Biosciences_Summary.pdf\"))
    base64 nil
    (\"attachment\"
     (modification-date . \"Wed, 18 Aug 2021 11:18:00 GMT\")
     (creation-date . \"Wed, 18 Aug 2021 11:18:00 GMT\")
     (size . \"55150\")
     (filename . \"UKRI-BBSRC_Scientific_Computing_in_the_Biosciences_Summary.pdf\"))
    \"UKRI-BBSRC_Scientific_Computing_in_the_Biosciences_Summary.pdf\" nil nil)
 (2 #<buffer  *mm*-226420>
    (\"text/plain\"
     (charset . \"us-ascii\"))
    quoted-printable nil
    (\"inline\")
    nil nil nil))
"

```

The first message was sent by an automated program, whereas the other messages I've tested are from 'humans', so perhaps the program is using the wrong structure for its attachments (but somehow I can see them if I click on them.)  but if the message is valid, could the filename be extracted using the name field of the application/octet-stream entry?

**To Reproduce**
I have the message (cannot share it publicly) with the above structure.  If you need to see it, let me know.

**Environment**
mu4e 1.6.1; Emacs 28.0.50; arch linux

**Checklist**
- [ X] you are running either the latest 1.6.x release, or a 1.7.x development release (otherwise, please upgrade).

**Comment by djcb** [comment]
As a work-around, there's the 'A' (mime-part action) which catches some more parts.

Perhaps you can attach the message in censored form? I.e. replace the
names / text etc., and the attachment can be empty blobs.

**Comment by sje30** [comment]
I'll look into mime-part action

in the meantime, I can email you the message off list --- is your email address in the commentary of the .el files still valid?

**Comment by djcb** [comment]
My address is still valid, but I prefer you censor the message and put
it here, so others can have a look as well.

**Comment by sje30** [comment]
sure, I can try, but how do I edit out the attachments etc?

**Comment by sje30** [comment]
ok, I had a go... editing the email file as plain text; deleted most of the base64 text, but hopefully youcan still see the structure... see http://damtp.cam.ac.uk/user/sje30/scrambled-email.txt

**Comment by danielfleischer** [comment]
Hi, I have the same issue. It seems the "attachment" alist is empty but file names can be extracted from the mime-type alist. From the few examples I'm seeing, the "name" alist is equal to the "filename" alist.

Please try and evaluate this code and tell me if it solves your issue, if so, I'll create a PR:

```elisp
(defun mu4e-view-save-attachments (&optional arg)
  "Save mime parts from current mu4e gnus view buffer.

When helm-mode is enabled provide completion on attachments and
possibility to mark candidates to save, otherwise completion on
attachments is done with `completing-read-multiple', in this case
use \",\" to separate candidate, completion is provided after
each \",\".

Note, currently this does not work well with file names
containing commas."
  (interactive "P")
  (cl-assert (and (eq major-mode 'mu4e-view-mode)
                  (derived-mode-p 'gnus-article-mode)))
  (let* ((parts (mu4e~view-gather-mime-parts))
         (handles '())
         (files '())
         (compfn (if (and (boundp 'helm-mode) helm-mode)
                     #'completing-read
                   ;; Fallback to `completing-read-multiple' with poor
                   ;; completion
                   #'completing-read-multiple))
        dir)
    (dolist (part parts)
      (print (dolist (p (seq-filter 'consp (cdr part))) (assoc 'name p)))
      (let ((fname (or (cdr (assoc 'filename (assoc "attachment" (cdr part))))
                       ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
                       ;;; These 3 lines are new
                       ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
                       (seq-find #'stringp
                                 (mapcar (lambda (item) (cdr (assoc 'name item)))
                                         (seq-filter 'listp (cdr part)))))))
        (when fname
          (push `(,fname . ,(cdr part)) handles)
          (push fname files))))
    (if files
        (progn
          (setq files (let ((helm-comp-read-use-marked t))
                        (funcall compfn "Save part(s): " files))
                dir (if arg (read-directory-name "Save to directory: ") mu4e-attachment-dir))
          (cl-loop for (f . h) in handles
                   when (member f files)
                   do (mm-save-part-to-file
                       h (let ((file (expand-file-name f dir)))
                           (if (file-exists-p file)
                               (let (newname (count 1))
                                 (while (and
                                         (setq newname
                                               (concat
                                                (file-name-sans-extension file)
                                                (format "(%s)" count)
                                                (file-name-extension file t)))
                                         (file-exists-p newname))
                                   (cl-incf count))
                                 newname)
                             file)))))
      (mu4e-message "No attached files found"))))
```

**Comment by thierryvolpiatto** [comment]
Daniel Fleischer ***@***.***> writes:

> Hi, I have the same issue. It seems the "attachment" alist is empty but file names can be extracted from the mime-type alist. From the few examples I'm seeing, the "name" alist is equal to
> the "filename" alist.
>
> Please try and evaluate this code and tell me if it solves your issue, if so, I'll create a PR:
>
>                        ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
>                        ;;; These 3 lines are new
>                        ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
>                        (seq-find #'stringp
>                                  (mapcar (lambda (item) (cdr (assoc 'name item)))
>                                          (seq-filter 'listp (cdr part)))))))

It seems you are running 3 loops whereas only one is needed:

1) seq-filter
2) mapcar on 1)
3) seq-find on 2)

Isn't the code below doing the same thing?

    (cl-loop for item in (cdr part)
             for name = (and (listp l) (assoc-default 'name item))
             thereis (and (stringp name) name)) 

-- 
Thierry

**Comment by danielfleischer** [comment]
Thank you for your feedback! I tried rewriting it using cl-loop as well; your code works (replace `l` with `item`).

**Comment by sje30** [comment]
thanks for this; my observation was that there are two tests:

1. (cdr (assoc 'filename (assoc "attachment" (cdr part))))

2. the new one just added

Is 1. needed -- or is 2 sufficient?

**Comment by danielfleischer** [comment]
Hi,


(1) is the existing logic, works in general. (2) is a fallback logic.
Being conservative and not knowing enough about how emails with
attachments usually look like, I won't claim (2) will work always. In
addition, or is lazy, stopping at the first non-nil condition so we
guarantee not to mess with what currently works.

**Comment by sje30** [comment]
Thanks, sounds sensible.

**Comment by danielfleischer** [comment]
But does it work for you?

**Comment by sje30** [comment]
yes, thank you!

**Comment by sje30** [comment]
p.s. https://github.com/sje30/emacs/blob/master/mu4e-view-save-all-attachments.el has the code I've adapted now to save all attachments, rather than ask for which attachments to save.

**Comment by danielfleischer** [comment]
Very nice; I'm using helm so I can choose all files with `M-a`.

**Comment by djcb** [comment]
Great; merged in the both release/1.6 and master. Closing.
